analyses/org.eclipse.tracecompass.incubator.inandout.core/src/org/eclipse/tracecompass/incubator/internal/inandout/core/analysis/InAndOutDataProviderFactory.java

```diff
@@ -245,17 +253,61 @@ private void applyConfiguration(ITmfTrace trace, ITmfConfiguration config, boole
         }
          // Apply configuration to any trace (no need to check trace type here)
         InAndOutAnalysisModule.create(config, trace, writeConfig);
+
+        // Save ID prefixes of children data providers
+        saveChildrenDescriptorPrefixes(trace);
     }
 
     private void removeConfiguration(ITmfTrace trace, ITmfConfiguration config) throws TmfConfigurationException {
         if (trace instanceof TmfExperiment) {
+            // Remove analysis modules in removeConfiguration per trace
             for (ITmfTrace tr : TmfTraceManager.getTraceSet(trace)) {
                 removeConfiguration(tr, config);
             }
+
+            // This will remove any dp instance (e.g. call stack dp)
+            removeChildrenDataProviders(trace, config);
+
             // only remove for traces in experiment
             return;
         }
+        // This will remove the analysis module
         InAndOutAnalysisModule.remove(config, trace);
+
+        // This will remove any dp instance (e.g. call stack dp)
+        removeChildrenDataProviders(trace, config);
+    }
+
+    private void saveChildrenDescriptorPrefixes(ITmfTrace trace) {
+        if (fDataProviderPrefixes == null) {
+            Predicate<IDataProviderDescriptor> predicate = desc -> {
+                ITmfConfiguration cfg = desc.getConfiguration();
+                return (cfg != null && cfg.getSourceTypeId().equals(SegmentSpecifierConfiguration.IN_AND_OUT_CONFIG_SOURCE_TYPE_ID));
+            };
+            Function<IDataProviderDescriptor, String> mapper = desc -> {
+                String id = desc.getId();
+                int index = id.indexOf(DataProviderConstants.CONFIG_SEPARATOR);
+                if (index > 0 && index < id.length()) {
+                    return id.substring(0, index);
+                }
+                return ""; //$NON-NLS-1$
+            };
+            fDataProviderPrefixes = DataProviderManager.getInstance().getAvailableProviders(trace)
+                    .stream()
+                    .filter(predicate)
+                    .map(mapper)
+                    .collect(Collectors.toList());
+        }
+    }
+
+    private void removeChildrenDataProviders(ITmfTrace trace, ITmfConfiguration config) {
+        List<String> dpPrefixes = fDataProviderPrefixes;
```

Comment:
I wonder, instead of computing the list of prefixes at creation, wouldn't it be simpler at removal to go through all available providers and remove any found whose id ends with +<configId> ?

Reply:
I had that initially. I changed it because I have to call getAvailableDataProviders() and find all relevant data providers for each trace and the experiment. I felt that was inefficient. WDYT?

Reply:
It seems the inefficiency at removal is the same as what would be avoided at creation, and we avoid having to extract and store the prefixes?

Reply:
The storing of prefixes is done only once.

Reply:
I'll change it back. I think added complexity is not really warrented.